### ISSUE TITLE
uploadMissing teardown causes concurrent action CancellationExceptions

### ISSUE THREAD

**Issue by werkt** [bazelbuild/bazel#10846]
### Description of the problem / feature request:

CancellationExceptions can occur on concurrently-waiting uploads on multiple actions that register as catastrophic build failures

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

This is a concurrency race, and difficult to reproduce. For two actions running through the remote spawn strategy, both uploading filesets in `uploadMissing` with overlapping digests, the future is shared between them for the overlap. When one action encounters an ExecutionException, it will cancel all of its uploads, which now causes a CancellationException to be thrown in the `upload.get()` for the second action. This CancellationException is uncaught and causes catastrophic build failure (see trace below).

Other sources for the CancellationException exist, but don't appear to occur here. The original action continues on, and if the ExecutionException was retriable, it is ignored. Only the secondary actions experience the uncaught exception and trigger the catastrophic build failure.

While handling CancellationException alongside the ExecutionException could resolve this, I believe the behavior of `uploadMissing` to assert ownership over the possibly shared upload is incorrect in triggering the cancel. The entire model may be insufficient to 'cancel the last upload of a digest' without substantial intervention.

### What operating system are you running Bazel on?

Ubuntu 18.04

### What's the output of `bazel info release`?

2.1.1

### Any other information, logs, or outputs that you want to share?

Trace from the exception:

```
Internal error thrown during build. Printing stack trace: java.lang.RuntimeException: Unrecoverable error while evaluating node 'ActionLookupData{actionLookupKey=//package:target BuildConfigurationValue.Key[<id>] false, actionIndex=3}' (requested by nodes 'ActionLookupData{actionLookupKey=//package:target BuildConfigurationValue.Key[<id>] false, actionIndex=9}')
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:515)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:399)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.util.concurrent.CancellationException: Task was cancelled.
        at com.google.common.util.concurrent.AbstractFuture.cancellationExceptionWithCause(AbstractFuture.java:1236)
        at com.google.common.util.concurrent.AbstractFuture.getDoneValue(AbstractFuture.java:524)
        at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:487)
        at com.google.common.util.concurrent.AbstractFuture$TrustedFuture.get(AbstractFuture.java:83)
        at com.google.devtools.build.lib.remote.RemoteExecutionCache.uploadMissing(RemoteExecutionCache.java:62)
        at com.google.devtools.build.lib.remote.RemoteExecutionCache.ensureInputsPresent(RemoteExecutionCache.java:126)
        at com.google.devtools.build.lib.remote.RemoteSpawnRunner.lambda$exec$0(RemoteSpawnRunner.java:287)
        at com.google.devtools.build.lib.remote.Retrier.execute(Retrier.java:237)
        at com.google.devtools.build.lib.remote.RemoteRetrier.execute(RemoteRetrier.java:115)
        at com.google.devtools.build.lib.remote.RemoteSpawnRunner.exec(RemoteSpawnRunner.java:278)
        at com.google.devtools.build.lib.exec.SpawnRunner.execAsync(SpawnRunner.java:229)
        at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:124)
        at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:89)
        at com.google.devtools.build.lib.actions.SpawnActionContext.beginExecution(SpawnActionContext.java:41)
        at com.google.devtools.build.lib.exec.ProxySpawnActionContext.beginExecution(ProxySpawnActionContext.java:60)
        at com.google.devtools.build.lib.rules.cpp.CppCompileAction.beginExecution(CppCompileAction.java:1397)
        at com.google.devtools.build.lib.actions.Action.execute(Action.java:124)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$4.execute(SkyframeActionExecutor.java:931)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.continueAction(SkyframeActionExecutor.java:1070)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:1041)
        at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:116)
        at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:77)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:608)
        at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:903)
        at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:297)
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:438)
        ... 4 more
```

**Comment by ulfjack** [comment]
What would be the source of the ExecutionException?

**Comment by ulfjack** [comment]
I may be able to get in a short-term fix to catch and handle CancellationException. I agree that this looks like a design flaw, and it may take more time to figure out what the right model is, so I think a short-term fix may be worthwhile.

**Comment by werkt** [comment]
> What would be the source of the ExecutionException?

Any exception which occurs during the execution of the `cacheProtocol.upload{Blob,File}`'s futured activity. Specifically for us, this will be ByteStreamUploader's AsyncUpload activity, which should be RemoteRetrier guarded and result in a retry-exhaustion exception, but could include an IO exception for the source stream, and regardless will be IOException wrapped (due to multiple layers of such) and be captured in RemoteSpawnRunner's outermost handler: `execLocallyAndUploadOrFail`

**Comment by werkt** [comment]
> I may be able to get in a short-term fix to catch and handle CancellationException. I agree that this looks like a design flaw, and it may take more time to figure out what the right model is, so I think a short-term fix may be worthwhile.

Short-term is better than nothing, we will know pretty quick if it was insufficient, but we have to circle back on a real solution for this. I also think it might just make sense *not* to try to cancel any of the other uploads for the quick fix.

In the long run, if it continues to operate on a shared in-progress upload registry, it needs something akin to a reference-counted future.

**Comment by ulfjack** [comment]
We are seeing a problem in recent Bazel versions that they *do not cancel* downloads when the build is interrupted with Ctrl-C, i.e., Bazel keeps downloading the output files in the background, but drops the received data. I'm wondering if it's related to the changes that happened here.

**Comment by coeuvre** [comment]
> We are seeing a problem in recent Bazel versions that they _do not cancel_ downloads when the build is interrupted with Ctrl-C, i.e., Bazel keeps downloading the output files in the background, but drops the received data. I'm wondering if it's related to the changes that happened here.

This behavior should be fixed by #12453.
